adjust_total_time before running on_complete callbacks

pedropb opened the pull request:
Closes #83

### What are you trying to accomplish?

We were using the `on_complete` callback to report the job total run time but we noticed that `total_time` only gets updated after the callbacks are called.

### What approach did you choose and why?

There were 3 options:
1. Move `adjust_total_time` so that it's called before the callbacks.
2. Copy `adjust_total_time` so that it's called before the callbacks and also right before logging the output summary. This could be useful in case the "total_time after callbacks" is valuable.
3. Publish a new event via `ActiveSupport::Notifications.instrument` instead. This would overlap with the callback functionality.

I decided for 1 because I think it's the one that makes the most sense.

### What reviewers should focus on?

- I'm not familiar with writing RubyGems so I added Timecop as a dependency without a group. If that's problematic I can try to come up with another solution to stub the passage of time.
- I would consider this a bugfix but if users are relying on the inaccurate `total_time` inside `on_complete` callbacks, this patch would be a breaking change.

Mangara commented on Gemfile:
```diff
@@ -27,3 +27,5 @@ gem "rake"
 # for unit testing optional sorbet support
 gem "sorbet-runtime"
+
+gem "timecop", "~> 0.9.4"
```
Could we use the [ActiveSupport time helpers](https://api.rubyonrails.org/v5.2.5/classes/ActiveSupport/Testing/TimeHelpers.html) instead? The gem already depends on parts of Rails, so that feels like a smaller step than a completely separate gem.

pedropb commented:
Latest pushed changes:
- Moved `adjust_total_time` from the `completed?` branch to the end of `iterate_with_enumerator`.
- Dropped TimeCop in favour of `ActiveSupport::Testing::TimeHelpers`.
- Consolidate the test files by merging the separate `callbacks_test.rb` logic into `iteration_test.rb`.

GustavoCaso commented on test/unit/iteration_test.rb:
```diff
@@ -219,6 +236,12 @@ def test_jobs_using_complex_but_serializable_cursor_will_not_raise
     work_one_job
   end
 
+  def test_jobs_using_on_complete_have_accurate_total_time
+    freeze_time do
+      JobThatCompletesAfter3Seconds.perform_now(->(job) { assert_equal(3, job.total_time) })
+    end
+  end
```
I would suggest simplifying the test:

```ruby
def test_jobs_using_on_complete_have_accurate_total_time
    freeze_time do
      job = JobThatCompletesAfter3Seconds.new
      job.perform_now
      assert_equal(3, job.total_time)
    end
  end
```

That way you do not need to pass the assertion around, which is a little weird 😄

GustavoCaso reviewed (commented):
Code LGTM. I would like it if you could modify the test file to remove the need to pass the assertions around. Apart from that this is ready to be merge 😄
